storage/innobase/dict/dict0mem.cc

```diff
@@ -312,14 +312,15 @@ dict_mem_table_add_col(
 
 	dict_mem_fill_column_struct(col, i, mtype, prtype, len);
 
-	switch (prtype & DATA_VERSIONED) {
-	case DATA_VERS_START:
-		ut_ad(!table->vers_start);
-		table->vers_start = i;
-		break;
-	case DATA_VERS_END:
-		ut_ad(!table->vers_end);
-		table->vers_end = i;
+	if ((prtype & DATA_UNVERSIONED) != DATA_UNVERSIONED) {
```

Comment:
> I think it would be better to avoid == or != in bitwise operations

I've tried to make code as easy to read as possible. I haven't checked whether current compiles optimize it or not. I'm sure they will in future. And I'm optimize code for reading, not for (possible) current compiles limitations.

`!(~prtype & DATA_UNVERSIONED)` will become `true` for unversioned fields. We're interested here only in `row_start` and `row_end`.

> Why was it necessary to change this code?

Looks like a general question. I answered it somewhere else. It wasn't strictly necessary. But IMO it makes code more clear.

Reply:
Sorry, I should have suggested
```c
if (~prtype & DATA_VERSIONED) {
```
This would hold whenever any of the `DATA_VERSIONED` bits are not set.
I would not rename the bitmask. Its current name is descriptive: it is related to versioning. Naming the bitmask for only one value (`DATA_UNVERSIONED`) would make the code harder to read for me. Usually, when a named flag (or group of flags) is set, the name of the bitmask should describe the situation. In this case, `DATA_VERSIONED` is an appropriate name, because the column would have to do something with system versioning.

And yes, there is a bad precedent: Setting `BTR_EXTERN_OWNER_FLAG` actually means ‘disowned’.